File: BinaryTree/java/leafList.java
import java.util.ArrayList;
import java.util.List;

class LeafList {
  public static List<Integer> leafList(TreeNode root) {
    List<Integer> leaves = new ArrayList<>();
    collect(root, leaves);
    return leaves;
  }

  private static void collect(TreeNode node, List<Integer> leaves) {
    if (node == null) {
      return;
    }
    if (node.left == null && node.right == null) {
      leaves.add(node.val);
      return;
    }
    collect(node.left, leaves);
    collect(node.right, leaves);
  }

  public static void run() {
    // intentionally empty
  }
}

File: BinaryTree/php/leafList.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/TreeNode.php';

function leafList(?TreeNode $root): array
{
    $leaves = [];
    collectLeaves($root, $leaves);
    return $leaves;
}

function collectLeaves(?TreeNode $node, array &$leaves): void
{
    if ($node === null) {
        return;
    }

    if ($node->left === null && $node->right === null) {
        $leaves[] = $node->val;
        return;
    }

    collectLeaves($node->left, $leaves);
    collectLeaves($node->right, $leaves);
}
